### ISSUE TITLE
[Duplicate Feature Request] Ability to save & export settings and custom queries

### ISSUE THREAD

**Issue by BZHuntShampu** [SpecterOps/BloodHound#18]
As this enhanced BloodHound project grows, surely more and more customization options will be developed.

It's be nice to be able to export/import those options along with settings and even customized queries in order to harmonize different instances of BloodHound CE.

**Comment by zinic** [comment]
We have this feature tracked internally and have plans to implement it. When we start moving on it we will provide updates to this issue.

Initial expectation is that we will support the existing custom query JSON format for import and export.

**Comment by StephenHinck** [comment]
We're tracking this in https://github.com/SpecterOps/BloodHound/issues/31

**Comment by StephenHinck** [comment]
On re-reading, this is slightly different than #31, so will leave open.

**Comment by n3rada** [comment]
I'm trying to figure out how the new community version works. I don't understand how to implement the `customqueries.json` that I had in the old bloodhound operation. If anyone knows...

**Comment by StephenHinck** [comment]
The queries presented within the UI are presently stored within bhce/packages/javascript/bh-shared-ui/src/commonSearches.tsx. We will be adding mechanisms to more easily import/export saved queries in the future.

**Comment by init5-SF** [comment]
> The queries presented within the UI are presently stored within bhce/packages/javascript/bh-shared-ui/src/commonSearches.tsx. We will be adding mechanisms to more easily import/export saved queries in the future.

Hello @StephenHinck , once this feature has been implemented, is it gonna be announced anywhere?
Thanks!

**Comment by StephenHinck** [comment]
Yes, all changes to features/functionality are included in release notes and this issue request would be closed at that time.

**Comment by MJLpl** [comment]
Is there the way to currently get customqueries.json mounted inside docker image?

**Comment by StephenHinck** [comment]
If you have existing customqueries, you could write a quick API integration to read those files and add them as saved queries within BHCE. See https://support.bloodhoundenterprise.io/hc/en-us/articles/11311053342619 for more details on how to use the API!

**Comment by init5-SF** [comment]
@StephenHinck hi,
Does bloodhound CE use the same syntax for queries as legacy bloodhound?
I've tried running ciphers manually on the CE version but they threw errors.

**Comment by StephenHinck** [comment]
It does. However, we still need to pull through support for a few cypher verbs, which is generally where you run into those issues today.

**Comment by exploide** [comment]
> you could write a quick API integration to read those files and add them as saved queries within BHCE

So I did this. I'm developing [bhcli](https://github.com/exploide/bhcli), a CLI tool for the BloodHound CE API. I recently added a [queries subcommand](https://github.com/exploide/bhcli#queries) to import and export custom queries.

```console
$ bhcli queries my-bloodhound-queries.json
INFO: Imported 12 custom queries.

$ bhcli queries --save queries-backup.json
INFO: Saved 12 queries to queries-backup.json
```

The `--save` option produces a comparably simple file format but import is also possible with legacy BloodHound's `customqueries.json` format.  Note that not everything from the latter might be compatible with BloodHound CE.

I hope this is useful for someone.

**Comment by init5-SF** [comment]
> > you could write a quick API integration to read those files and add them as saved queries within BHCE
> 
> So I did this. I'm developing [bloodhound-cli](https://github.com/exploide/bloodhound-cli), a CLI tool for the BloodHound CE API. I recently added a [queries subcommand](https://github.com/exploide/bloodhound-cli#queries) to import and export custom queries.
> 
> ```
> $ bhcli queries my-bloodhound-queries.json
> INFO: Imported 12 custom queries.
> 
> $ bhcli queries --save queries-backup.json
> INFO: Saved 12 queries to queries-backup.json
> ```
> 
> The `--save` option produces a comparably simple file format but import is also possible with legacy BloodHound's `customqueries.json` format. Note that not everything from the latter might be compatible with BloodHound CE.
> 
> I hope this is useful for someone.

Hello @exploide, this looks very promising!
Can it actually import _any_ type of custom queries from legacy BloodHound to the CE version?🤔 (i.e. AD CS queries, etc.)
If not, what are the queries that aren't supported yet?

**Comment by exploide** [comment]
> Hello @exploide, this looks very promising!
Can it actually import any type of custom queries from legacy BloodHound to the CE version?🤔 (i.e. AD CS queries, etc.)
If not, what are the queries that aren't supported yet?

It doesn't do any conversion, it just automates the import of queries. If the query uses legacy properties which changed in BHCE, the query won't work (however, the import would probably look successful ). For example the way high value / tier 0 nodes are tagged changed between BloodHound legacy and CE, so this will not be compatible. The ADCS stuff also got completely reinvented.
Furthermore, BloodHound legacy had this optional node selection feature, where you needed to select a node from a selection widget before the query actually runs. This does not exist in BHCE and such queries are not imported by `bhcli`.

**Comment by init5-SF** [comment]
> > Hello @exploide, this looks very promising!
> > Can it actually import any type of custom queries from legacy BloodHound to the CE version?🤔 (i.e. AD CS queries, etc.)
> > If not, what are the queries that aren't supported yet?
> 
> It doesn't do any conversion, it just automates the import of queries. If the query uses legacy properties which changed in BHCE, the query won't work (however, the import would probably look successful ). For example the way high value / tier 0 nodes are tagged changed between BloodHound legacy and CE, so this will not be compatible. The ADCS stuff also got completely reinvented. Furthermore, BloodHound legacy had this optional node selection feature, where you needed to select a node from a selection widget before the query actually runs. This does not exist in BHCE and such queries are not imported by `bhcli`.

Huge bummer. :(
Converting legacy queries to the new CE queries format is basically what I've been seeking desperately ever since CE version was released.
If you could add such functionality to `bhcli` this would be a total lifesaver for a larger number of BH users!

**Comment by exploide** [comment]
> Huge bummer. :(
Converting legacy queries to the new CE queries format is basically what I've been seeking desperately ever since CE version was released.
If you could add such functionality to bhcli this would be a total lifesaver for a larger number of BH users!

I understand your problem but I fear this is no easy task. And while it might be possible to have some kind of conversion script for typical cases (which would probably fail in a lot of edge cases) I don't really see this in scope of `bhcli`. This would rather be a script on its own. Though haven't seen any of this yet.
I don't know if you deal with hundreds of queries, but I guess many users will just adapt their most important queries by hand.

**Comment by StephenHinck** [comment]
This was released today with https://github.com/SpecterOps/BloodHound/releases/tag/v8.2.0

Full release notes: https://bloodhound.specterops.io/resources/release-notes/v8-2-0

Thanks for the input!
